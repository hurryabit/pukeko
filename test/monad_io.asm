g_declare_cafs gm$cons_0_0, dict$Ord$Int, dict$Ring$Int, dict$Monad$IO, main
g_declare_main main

g_globstart gm$cons_0_0, 0
g_updcons 0, 0, 1
g_return

g_globstart gm$cons_0_2, 2
g_updcons 0, 2, 1
g_return

g_globstart gm$cons_0_4, 4
g_updcons 0, 4, 1
g_return

g_globstart op$ge, 1
g_push 0
g_eval
g_uncons 4
g_pop 2
g_update 3
g_pop 2
g_unwind

g_globstart op$g, 1
g_push 0
g_eval
g_uncons 4
g_pop 3
g_update 2
g_pop 1
g_unwind

g_globstart op$m, 1
g_push 0
g_eval
g_uncons 4
g_pop 2
g_update 3
g_pop 2
g_unwind

g_globstart gm$lt, 2
g_push 1
g_eval
g_push 1
g_eval
g_les
g_update 3
g_pop 2
g_return

g_globstart gm$le, 2
g_push 1
g_eval
g_push 1
g_eval
g_leq
g_update 3
g_pop 2
g_return

g_globstart gm$ge, 2
g_push 1
g_eval
g_push 1
g_eval
g_geq
g_update 3
g_pop 2
g_return

g_globstart gm$gt, 2
g_push 1
g_eval
g_push 1
g_eval
g_gtr
g_update 3
g_pop 2
g_return

g_globstart dict$Ord$Int, 0
g_pushglobal gm$lt, 2
g_pushglobal gm$le, 2
g_pushglobal gm$ge, 2
g_pushglobal gm$gt, 2
g_push 0
g_push 2
g_push 4
g_push 6
g_updcons 0, 4, 5
g_pop 4
g_return

g_globstart gm$neg, 1
g_eval
g_neg
g_update 1
g_return

g_globstart gm$add, 2
g_push 1
g_eval
g_push 1
g_eval
g_add
g_update 3
g_pop 2
g_return

g_globstart gm$sub, 2
g_push 1
g_eval
g_push 1
g_eval
g_sub
g_update 3
g_pop 2
g_return

g_globstart gm$mul, 2
g_push 1
g_eval
g_push 1
g_eval
g_mul
g_update 3
g_pop 2
g_return

g_globstart dict$Ring$Int, 0
g_pushglobal gm$neg, 1
g_pushglobal gm$add, 2
g_pushglobal gm$sub, 2
g_pushglobal gm$mul, 2
g_push 0
g_push 2
g_push 4
g_push 6
g_updcons 0, 4, 5
g_pop 4
g_return

g_globstart pure, 1
g_push 0
g_eval
g_uncons 2
g_update 3
g_pop 2
g_unwind

g_globstart op$gge, 1
g_push 0
g_eval
g_uncons 2
g_pop 1
g_update 2
g_pop 1
g_unwind

g_globstart op$s$ll1, 2
g_update 2
g_pop 1
g_unwind

g_globstart op$s, 3
g_push 2
g_pushglobal op$s$ll1, 2
g_mkap 1
g_push 2
g_push 2
g_pushglobal op$gge, 1
g_mkap 1
g_updap 2, 4
g_pop 3
g_unwind

g_globstart when, 3
g_push 1
g_eval
g_jumpcase .0, .1
g_label .0
g_pop 1
g_pushglobal gm$cons_0_0, 0
g_push 1
g_pushglobal pure, 1
g_mkap 1
g_updap 1, 4
g_pop 3
g_unwind
g_label .1
g_pop 3
g_update 1
g_unwind
g_label .2

g_globstart gm$return, 2
g_updcons 0, 2, 1
g_return

g_globstart gm$bind, 3
g_push 2
g_push 1
g_mkap 1
g_eval
g_uncons 2
g_push 3
g_updap 2, 4
g_pop 3
g_unwind

g_globstart dict$Monad$IO, 0
g_pushglobal gm$return, 2
g_pushglobal gm$bind, 3
g_push 0
g_push 2
g_updcons 0, 2, 3
g_pop 2
g_return

g_globstart gm$print, 2
g_eval
g_print
g_cons 0, 0
g_updcons 0, 2, 1
g_return

g_globstart gm$input, 1
g_input
g_updcons 0, 2, 1
g_return

g_globstart count_down, 1
g_pushint 1
g_push 1
g_pushglobal dict$Ring$Int, 0
g_pushglobal op$m, 1
g_mkap 1
g_mkap 2
g_pushglobal count_down, 1
g_mkap 1
g_push 1
g_pushglobal gm$print, 2
g_mkap 1
g_pushglobal dict$Monad$IO, 0
g_pushglobal op$s, 3
g_mkap 1
g_mkap 2
g_pushint 0
g_push 2
g_pushglobal dict$Ord$Int, 0
g_pushglobal op$ge, 1
g_mkap 1
g_mkap 2
g_pushglobal dict$Monad$IO, 0
g_pushglobal when, 3
g_mkap 1
g_updap 2, 2
g_pop 1
g_unwind

g_globstart repeat, 3
g_push 2
g_pushint 1
g_push 3
g_pushglobal dict$Ring$Int, 0
g_pushglobal op$m, 1
g_mkap 1
g_mkap 2
g_push 2
g_pushglobal repeat, 3
g_mkap 1
g_mkap 2
g_push 3
g_push 2
g_pushglobal op$s, 3
g_mkap 1
g_mkap 2
g_pushint 0
g_push 3
g_pushglobal dict$Ord$Int, 0
g_pushglobal op$g, 1
g_mkap 1
g_mkap 2
g_push 2
g_pushglobal when, 3
g_mkap 1
g_updap 2, 4
g_pop 3
g_unwind

g_globstart main$ll1, 2
g_push 1
g_pushglobal count_down, 1
g_mkap 1
g_push 1
g_pushglobal dict$Monad$IO, 0
g_pushglobal repeat, 3
g_mkap 1
g_updap 2, 3
g_pop 2
g_unwind

g_globstart main$ll2, 1
g_push 0
g_pushglobal main$ll1, 2
g_mkap 1
g_pushglobal gm$input, 1
g_pushglobal dict$Monad$IO, 0
g_pushglobal op$gge, 1
g_mkap 1
g_updap 2, 2
g_pop 1
g_unwind

g_globstart main, 0
g_pushglobal main$ll2, 1
g_pushglobal gm$input, 1
g_pushglobal dict$Monad$IO, 0
g_pushglobal op$gge, 1
g_mkap 1
g_updap 2, 1
g_unwind
