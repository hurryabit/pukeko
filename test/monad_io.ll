external (-) = "sub"
external (>=) = "ge"
external (>) = "gt"
external return = "return"
external print = "print"
external input = "input"
external (>>=) = "bind"
let (;ll1) = fun m2 x -> m2
let (;) = fun m1 m2 -> (>>=) m1 ((;ll1) m2)
let when =
      fun p m ->
        match p with
        | False -> return Unit
        | True -> m
let count_down =
      fun k -> when ((>=) k 0) ((;) (print k) (count_down ((-) k 1)))
let repeat_m =
      fun k m -> when ((>) k 0) ((;) m (repeat_m ((-) k 1) m))
let main$ll2 = fun k n -> repeat_m k (count_down n)
let main$ll1 = fun k -> (>>=) input (main$ll2 k)
let main = (>>=) input main$ll1
