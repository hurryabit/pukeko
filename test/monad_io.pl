data Unit =
       | Unit
data Bool =
       | False
       | True
data Pair a b =
       | Pair a b
data Option a =
       | None
       | Some a
data Choice a b =
       | First a
       | Second b
data Eq a =
       | Dict$Eq (a -> a -> Bool)
data Ord a =
       | Dict$Ord (a -> a -> Bool) (a -> a -> Bool) (a -> a -> Bool) (a -> a -> Bool)
data Monoid m =
       | Dict$Monoid m (m -> m -> m)
data Ring a =
       | Dict$Ring (a -> a) (a -> a -> a) (a -> a -> a) (a -> a -> a)
data Char
data Foldable t =
       | Dict$Foldable (∀a b. (a -> b -> b) -> b -> t a -> b) (∀a b. (b -> a -> b) -> b -> t a -> b)
data Functor f =
       | Dict$Functor (∀a b. (a -> b) -> f a -> f b)
data List a =
       | Nil
       | Cons a (List a)
data Monad m =
       | Dict$Monad (∀a. a -> m a) (∀a b. m a -> (a -> m b) -> m b)
data World =
       | World
data IO a = World -> Pair a World
external lt_int : Int -> Int -> Bool = "lt"
external le_int : Int -> Int -> Bool = "le"
external ge_int : Int -> Int -> Bool = "ge"
external gt_int : Int -> Int -> Bool = "gt"
external neg_int : Int -> Int = "neg"
external add_int : Int -> Int -> Int = "add"
external sub_int : Int -> Int -> Int = "sub"
external mul_int : Int -> Int -> Int = "mul"
external seq : ∀a b. a -> b -> b = "seq"
external puti : Int -> Unit = "puti"
external geti : Unit -> Int = "geti"
dict$Ord$Int : Ord Int =
  let ge : Int -> Int -> Bool = ge_int
  and gt : Int -> Int -> Bool = gt_int
  and le : Int -> Int -> Bool = le_int
  and lt : Int -> Int -> Bool = lt_int
  in
  Dict$Ord @Int ge gt le lt
dict$Ring$Int : Ring Int =
  let neg : Int -> Int = neg_int
  and add : Int -> Int -> Int = add_int
  and sub : Int -> Int -> Int = sub_int
  and mul : Int -> Int -> Int = mul_int
  in
  Dict$Ring @Int neg add sub mul
dict$Monad$IO : Monad IO =
  let pure : ∀a. a -> IO a = dict$Monad$IO$ll2
  and bind : ∀a b. IO a -> (a -> IO b) -> IO b = dict$Monad$IO$ll4
  in
  Dict$Monad @IO pure bind
input : IO Int = io$ll2 @Unit @Int geti Unit
main : IO Unit =
  bind$ll1 @IO dict$Monad$IO @Int @Unit input main$ll2
ge$ll1 : ∀a. Ord a -> a -> a -> Bool =
  fun @a (dict : Ord a) ->
    match dict with
    | Dict$Ord @a ge _ _ _ -> ge
gt$ll1 : ∀a. Ord a -> a -> a -> Bool =
  fun @a (dict : Ord a) ->
    match dict with
    | Dict$Ord @a _ gt _ _ -> gt
sub$ll1 : ∀a. Ring a -> a -> a -> a =
  fun @a (dict : Ring a) ->
    match dict with
    | Dict$Ring @a _ _ sub _ -> sub
pure$ll1 : ∀m. Monad m -> (∀a. a -> m a) =
  fun @m (dict : Monad m) ->
    match dict with
    | Dict$Monad @m pure _ -> pure
bind$ll1 : ∀m. Monad m -> (∀a b. m a -> (a -> m b) -> m b) =
  fun @m (dict : Monad m) ->
    match dict with
    | Dict$Monad @m _ bind -> bind
semi$ll1 : ∀a m. m a -> Unit -> m a =
  fun @a @m (m2 : m a) (x : Unit) -> m2
semi$ll2 : ∀a m. Monad m -> m Unit -> m a -> m a =
  fun @a @m (dict$Monad$m : Monad m) (m1 : m Unit) (m2 : m a) ->
    bind$ll1 @m dict$Monad$m @Unit @a m1 (semi$ll1 @a @m m2)
when$ll1 : ∀m. Monad m -> Bool -> m Unit -> m Unit =
  fun @m (dict$Monad$m : Monad m) (p : Bool) (m : m Unit) ->
    match p with
    | False -> pure$ll1 @m dict$Monad$m @Unit Unit
    | True -> m
dict$Monad$IO$ll1 : ∀a. a -> World -> Pair a World =
  fun @a -> Pair @a @World
dict$Monad$IO$ll2 : ∀a. a -> IO a =
  fun @a (x : a) -> coerce @(_ -> IO) (dict$Monad$IO$ll1 @a x)
dict$Monad$IO$ll3 : ∀a b. IO a -> (a -> IO b) -> World -> Pair b World =
  fun @a @b (mx : IO a) (f : a -> IO b) (world0 : World) ->
    match coerce @(IO -> _) mx world0 with
    | Pair @a @World x world1 -> coerce @(IO -> _) (f x) world1
dict$Monad$IO$ll4 : ∀a b. IO a -> (a -> IO b) -> IO b =
  fun @a @b (mx : IO a) (f : a -> IO b) ->
    coerce @(_ -> IO) (dict$Monad$IO$ll3 @a @b mx f)
io$ll1 : ∀a b. (a -> b) -> a -> World -> Pair b World =
  fun @a @b (f : a -> b) (x : a) (world : World) ->
    let y : b = f x in
    seq @b @(Pair b World) y (Pair @b @World y world)
io$ll2 : ∀a b. (a -> b) -> a -> IO b =
  fun @a @b (f : a -> b) (x : a) ->
    coerce @(_ -> IO) (io$ll1 @a @b f x)
print$ll1 : Int -> IO Unit = io$ll2 @Int @Unit puti
count_down$ll1 : Int -> IO Unit =
  fun (k : Int) ->
    when$ll1 @IO dict$Monad$IO (ge$ll1 @Int dict$Ord$Int k 0) (semi$ll2 @Unit @IO dict$Monad$IO (print$ll1 k) (count_down$ll1 (sub$ll1 @Int dict$Ring$Int k 1)))
repeat$ll1 : ∀m. Monad m -> Int -> m Unit -> m Unit =
  fun @m (dict$Monad$m : Monad m) (k : Int) (m : m Unit) ->
    when$ll1 @m dict$Monad$m (gt$ll1 @Int dict$Ord$Int k 0) (semi$ll2 @Unit @m dict$Monad$m m (repeat$ll1 @m dict$Monad$m (sub$ll1 @Int dict$Ring$Int k 1) m))
main$ll1 : Int -> Int -> IO Unit =
  fun (k : Int) (n : Int) ->
    repeat$ll1 @IO dict$Monad$IO k (count_down$ll1 n)
main$ll2 : Int -> IO Unit =
  fun (k : Int) ->
    bind$ll1 @IO dict$Monad$IO @Int @Unit input (main$ll1 k)
