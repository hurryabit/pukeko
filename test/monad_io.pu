val when : Bool -> IO Unit -> IO Unit
let when p m = if p then m else return Unit

val count_down : Int -> IO Unit
let rec count_down k = when (k >= 0) (print k; count_down (k-1))

val repeat_m : Int -> IO Unit -> IO Unit
let rec repeat_m k m = when (k > 0) (m; repeat_m (k-1) m)

val main : IO Unit
let main = repeat_m 3 (count_down 2)
