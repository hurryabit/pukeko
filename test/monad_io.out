2
1
0
2
1
0
2
1
0

Steps  :        15172
Allocs :          397          (Checks:          314)
GC runs:            0
        0.00 real         0.00 user         0.00 sys
